namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala

```diff
@@ -212,27 +279,32 @@ class ThriftNamerInterface(
 
   private[this] val addrObserverCacheMu = new {}
   private[this] var addrObserverCache: Map[Path, AddrObserver] = Map.empty
-  private[this] def cacheBoundAddrs(nt: NameTree[Name.Bound]): Unit = nt match {
-    case NameTree.Neg =>
-    case NameTree.Empty =>
-    case NameTree.Fail =>
-    case NameTree.Alt(trees@_*) =>
-      trees.foreach(cacheBoundAddrs)
-    case NameTree.Union(trees@_*) =>
-      trees.foreach(w => cacheBoundAddrs(w.tree))
-    case NameTree.Leaf(bound) =>
-      bound.id match {
-        case id: Path =>
-          addrObserverCacheMu.synchronized {
-            if (!addrObserverCache.contains(id)) {
-              addrObserverCache += (id -> AddrObserver(bound.addr, stamper))
-            }
+  private[this] def observeAddr(ns: String, id: Path): AddrObserver =
+    addrObserverCacheMu.synchronized {
+      addrObserverCache.get(id) match {
+        case Some(obs) =>
+          Trace.recordBinary("namerd.srv/addr.cached", true)
+          obs
+        case None =>
+          Trace.recordBinary("namerd.srv/addr.cached", false)
+          val addr = getBind(ns, Dtab.empty, id).run.flatMap {
+            case Activity.Failed(e) => Var.value(Resolution.Resolved(Addr.Failed(e)))
+            case Activity.Pending => Var.value(Resolution.Resolved(Addr.Pending))
+            case Activity.Ok(NameTree.Leaf(bound)) if bound.id == id =>
+              bound.addr.map(Resolution.Resolved(_))
+            case _ => Var.value(Resolution.Released)
```

Comment:
For clarity can we refine this pattern?  This is really `case Activity.Ok(_)`, right?  And moreover, we expect this to be triggered when it goes to `Activity.Ok(NameTree.Neg)`.  Any other value would be a violation of the assumption that the input is already a bound name, I think.